### PR TITLE
[RLlib] Rnnsac Issues

### LINKED ISSUE #22693: [Feature] [RLLib] Remove RNNSAC
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

RNNSAC is badly broken.  I lost a lot of time trying to use it.  It would be an improvement to RLLib to remove it as it is not maintained.   And obviously no one has ever used it since there are no bug reports.

I found many bugs and I'm sure there are many more.

`RNNSACTrainer.validate_config()` modifies the input `config["replay_sequence_length"]`.  But config validation is called twice in the grand scheme of things and it crashes the second time it's called.

In rnn_torch_sac_policy.py:`action_distribution_fn()` calls` model.get_q_values()` (twice if twin_q) but without adding the final `actions` argument `input_dict["actions"]`.  Somehow `get_q_values()` accepts a missing `actions` and uses `None` instead.  But that crashed in my use case.  I had to pass the actions within `action_distribution_fn()`, which are anyway available as `input_dict["actions"]`.

`RNNSACTorchModel._get_q_value()` is called twice when twin_q is used.  But the input argument `model_out` is modified by the function and it crashes the second time it's called with a now modified `model_out` input.

Also I had to replace `RNNSACTorchModel.get_policy_output()` with `SACTorchModel.get_policy_output()`.  The SAC version does a lot of preprocessing to the argument `model_out` which was certainly needed for my use case.

Something else which may be completely irrelevant is that `ComplexInputNetwork.__init__()` calls `TorchModelV2.__init__(..., self.original_space, ...)`, but I changed it to `obs_space` instead of `original_space`.  I honestly can't remember why I did that; I'm not sure if I needed to change that to get RNNSAC working.   But it might break RLLib elsewhere to do that.   I'm just mentioning this so you can take a look and double check if the current `TorchModelV2.__init__(..., self.original_space, ...)` is correct.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

A couple of things about our RNNSAC implementation are broken.
This PR is a collection of smaller code changes and refractors that aim to aid with this.

Closes #22693 

## Related issue number

Issue #22693

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/sac/sac_torch_model.py [hunk 1]
```diff
@@ -284,26 +289,41 @@ def get_policy_output(self, model_out: TensorType) -> TensorType:
...
         """
-        # Model outs may come as original Tuple observations, concat them
-        # here if this is the case.
-        if isinstance(self.action_model.obs_space, Box):
-            if isinstance(model_out, (list, tuple)):
-                model_out = torch.cat(model_out, dim=-1)
-            elif isinstance(model_out, dict):
-                model_out = torch.cat(
+
+        def concat_obs_if_necessary(obs):
```
Comment (gjoliver) [inline rllib/agents/sac/sac_torch_model.py:301]: help type-annotate obs here?

**Review by gjoliver — COMMENTED** [review]
thanks, looks good.
this also makes me curious if rnnsac has ever worked ...

File: rllib/agents/sac/sac_tf_model.py [hunk 2]
```diff
@@ -120,6 +121,10 @@ def __init__(
                 target_entropy = -np.prod(action_space.shape)
         self.target_entropy = target_entropy
 
+        self.flattened_action_model_space = flatten_space(
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/agents/sac/sac_tf_model.py:124]: Leftover from another approach related to user's issue. Thanks!

File: rllib/agents/sac/sac_tf_model.py [hunk 3]
```diff
@@ -276,26 +286,41 @@ def get_policy_output(self, model_out: TensorType) -> TensorType:
...
         """
-        # Model outs may come as original Tuple/Dict observations, concat them
-        # here if this is the case.
-        if isinstance(self.action_model.obs_space, Box):
-            if isinstance(model_out, (list, tuple)):
-                model_out = tf.concat(model_out, axis=-1)
-            elif isinstance(model_out, dict):
-                model_out = tf.concat(
+
+        def concat_obs_if_necessary(obs):
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/agents/sac/sac_tf_model.py:298]: `obs` might also be a nested dict of some sort.
This happens if we don't have a base model prepended to the actor.
Then it seems like our solution is to simply hand the observation to the actor.
The observation at that point might still be nested observations!
I don't know if this is the coolest solution, but at this point I don't know if I should invest the time to find the coolest solution for this algorith.
What do you say?

[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/sac/sac_tf_model.py]: Really, we support that? good to know :)
It's fine. I was just curious how this works.

**Review by gjoliver — APPROVED** [review]

**Comment by gjoliver** [comment]
a few unit tests to fix though.

File: rllib/agents/sac/rnnsac.py [hunk 4]
```diff
@@ -47,16 +47,26 @@ def validate_config(self, config: TrainerConfigDict) -> None:
...
+                "config['replay_sequence_length'] blank to avoid this error."
             )
-        # Add the `burn_in` to the Model's max_seq_len.
         # Set the replay sequence length to the max_seq_len of the model.
-        config["replay_buffer_config"]["replay_sequence_length"] = (
-            config["burn_in"] + config["model"]["max_seq_len"]
-        )
+        config["replay_buffer_config"][
+            "replay_sequence_length"
+        ] = replay_sequence_length
```
Comment (gjoliver) [inline rllib/agents/sac/rnnsac.py:69]: @ArturNiederfahrenhorst @sven1977 I have a nit about this pattern that I am seeing a lot throughout RLlib's codebase.
I think we should treat config as a "const" obj. This is what users tell us, and likely what we should record with our checkpoints etc. modifying this in-place feels pretty confusing.
if we need to derive a control bit from some user configured inputs, we can create an internal only local configuration, but that should be part of the implementation details, not getting reflected back at the original cfg.

File: rllib/agents/sac/sac_tf_model.py [hunk 5]
```diff
@@ -263,11 +263,16 @@ def _get_q_value(self, model_out, actions, net):
         # training).
         input_dict["is_training"] = True
 
-        out, _ = net(input_dict, [], None)
-        return out
+        return net(input_dict, [], None)
```
Comment (gjoliver) [inline rllib/agents/sac/sac_tf_model.py:266]: where are you using this additional output?? sorry, it's a bit hard to see.

**Pull request merged**